# Exploring eBay Car Sales Data


###  In this project, We'll work a data set of used cars from eBay kleinanzeigan,  a classified section of the German eBay website.

#####    It's just sampled 50,000 data points from the full dataset, And here is the columns info:
dateCrawled - When this ad was first crawled. All field-values are taken from this date.
name - Name of the car.
seller - Whether the seller is private or a dealer.
offerType - The type of listing
price - The price on the ad to sell the car.
abtest - Whether the listing is included in an A/B test.
vehicleType - The vehicle Type.
yearOfRegistration - The year in which which year the car was first registered.
gearbox - The transmission type.
powerPS - The power of the car in PS.
model - The car model name.
kilometer - How many kilometers the car has driven.
monthOfRegistration - The month in which which year the car was first registered.
fuelType - What type of fuel the car uses.
brand - The brand of the car.
notRepairedDamage - If the car has a damage which is not yet repaired.
dateCreated - The date on which the eBay listing was created.
nrOfPictures - The number of pictures in the ad.
postalCode - The postal code for the location of the vehicle.
lastSeenOnline - When the crawler saw this ad last online.
##### First of all, We import the Numpy & Pandas libraries. Then, You maybe noticed that we used encoding Argument with 'pd.read_csv'; because it is decoded by "Latin-1".
##### After that, We used DataFrame.info() and DataFrame.head() to undrastand more about our dataframe.

In [1]:
import pandas as pd
import numpy as np
auto=pd.read_csv('autos.csv',encoding='Latin-1')
auto

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [2]:
auto.info()
auto.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

##### in coming cammands, Beacause of not adhering The Column Labels of snakecase, We should change them manually by copying the dataframe columns label in a new series, and copy it's content and finally change it with the intended needs to be in Snakecase instead of CamelCase.

In [3]:
column_name=auto.columns.copy()
column_name

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [2]:
auto.columns=['date_Crawled', 'name', 'seller', 'offer_Type', 'price', 'ab_test',
       'vehicle_Type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_Type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_Of_Pictures', 'postal_Code',
       'last_Seen']
auto.head()

date_Crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_Type   price  ab_test vehicle_Type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_Type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_Of_Pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_Code            last_Seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [5]:
auto.describe(include='all')

date_Crawled         name  seller offer_Type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_Type  registration_year gear_box      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_Type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_Of_Pictures   postal_Code            last_Seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [6]:
auto['odometer'].value_counts(dropna=False)

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

### Clean the Data
now It's time to valuate the columns and their entries by checking them one by one and see the correlations if is helpful.We catagorize them to 4 subcategories. first, which does not need any change, second which that have mostly one value that are candidates to be dropped, third which need more investigation and finally the last one which numeric data stored as text that needs to be cleaned.
We just talk about the last three ones.
the columns date_Crawled, ad_created,last_Seen are the date type which could be transform to numeric type, so they're in fourth sub.
And column odometer,price, could be easily transform to numeric column, so in fourth one.
The column nr_Of_Pictures does not have any valuable content; and the columns seller and offer_type have the same information and one is additional, So in the second one.
The Columns gear gear_box, power_PS, unrepaired_damage, fuel_Type, registration_month , model, because of some reasons need more investigation. So in the third one.


In [3]:
auto['price']=auto['price'].str.replace('$','')
auto['price']=auto['price'].str.replace(',','').astype(float)
auto['odometer']=auto['odometer'].str.replace('km','')
auto['odometer']=auto['odometer'].str.replace(',','').astype(float)
auto.rename({'odometer':'odometer_km'},axis=1,inplace=True)

In [8]:
auto.describe(include='all')

date_Crawled         name  seller offer_Type         price  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       ab_test vehicle_Type  registration_year gear_box      power_PS  model  \
count    50000        44905       50000.000000    47320  50000.000000  47242   
unique       2            8                NaN        2           NaN    245   
top       test    limousine                NaN  manuell           NaN   golf   
freq     25756        12859                NaN    36993           NaN   4024   
mean       NaN          NaN        2005.073280      NaN    116.355920    NaN   
std        NaN          NaN         105.712813      NaN    209.216627    NaN   
min        NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%        NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%        NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%        NaN          NaN        2008.000000      NaN    150.000000    NaN   
max        NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_Type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_Of_Pictures   postal_Code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_Seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [4]:
auto['odometer_km'].value_counts(dropna=False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In [5]:
auto['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [13]:
auto['odometer_km'].unique().shape

(13,)

In [15]:
auto['price'].value_counts(dropna=False).head(15)


0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
1000.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
999.0      434
750.0      433
900.0      420
650.0      419
850.0      410
Name: price, dtype: int64

In [16]:
auto['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [4]:
auto=auto[auto['odometer_km'].between(87500,150000)]
auto.describe(include='all')

date_Crawled      name  seller offer_Type         price  \
count                 41520     41520   41520      41520  4.152000e+04   
unique                40294     32335       2          2           NaN   
top     2016-04-02 15:49:30  BMW_316i  privat    Angebot           NaN   
freq                      3        73   41519      41519           NaN   
mean                    NaN       NaN     NaN        NaN  8.237172e+03   
std                     NaN       NaN     NaN        NaN  5.072893e+05   
min                     NaN       NaN     NaN        NaN  0.000000e+00   
25%                     NaN       NaN     NaN        NaN  9.990000e+02   
50%                     NaN       NaN     NaN        NaN  2.450000e+03   
75%                     NaN       NaN     NaN        NaN  5.500000e+03   
max                     NaN       NaN     NaN        NaN  1.000000e+08   

       ab_test vehicle_Type  registration_year gear_box      power_PS  model  \
count    41520        37209       41520.000000    39421  41520.000000  39310   
unique       2            8                NaN        2           NaN    242   
top       test    limousine                NaN  manuell           NaN   golf   
freq     21359        10912                NaN    31198           NaN   3407   
mean       NaN          NaN        2002.820111      NaN    114.490390    NaN   
std        NaN          NaN          35.023528      NaN    192.723287    NaN   
min        NaN          NaN        1910.000000      NaN      0.000000    NaN   
25%        NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%        NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%        NaN          NaN        2007.000000      NaN    147.000000    NaN   
max        NaN          NaN        9000.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_Type       brand  \
count    41520.000000        41520.000000     37762       41520   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     24355        9097   
mean    141736.030829            5.718015       NaN         NaN   
std      17102.004255            3.731914       NaN         NaN   
min      90000.000000            0.000000       NaN         NaN   
25%     150000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_Of_Pictures   postal_Code  \
count              32974                41520         41520.0  41520.000000   
unique                 2                   70             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               28475                 1593             NaN           NaN   
mean                 NaN                  NaN             0.0  50590.652192   
std                  NaN                  NaN             0.0  25622.203369   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30179.000000   
50%                  NaN                  NaN             0.0  49328.000000   
75%                  NaN                  NaN             0.0  71083.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_Seen  
count                 41520  
unique                34391  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [5]:
auto=auto[auto['price'].between(0,16350)]
auto

date_Crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
32     2016-03-20 05:03:03                               Corsa_mit_TÜV_5.2016   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34     2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
36     2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
...                    ...                                                ...   
49963  2016-03-26 19:52:59            Mercedes_Benz_B_200_CDI_Special_Edition   
49964  2016-03-10 16:46:43                                 Audi_2_7_TDI_AVANT   
49965  2016-03-11 10:53:51          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49967  2016-03-12 11:50:17                      VW_Passat_2_0_TDI_comfortline   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_F

In [20]:
auto.describe(include='all')

date_Crawled      name  seller offer_Type         price  \
count                 39964     39964   39964      39964  39964.000000   
unique                38823     31112       2          2           NaN   
top     2016-03-23 18:39:34  BMW_316i  privat    Angebot           NaN   
freq                      3        73   39963      39963           NaN   
mean                    NaN       NaN     NaN        NaN   3558.803899   
std                     NaN       NaN     NaN        NaN   3536.038723   
min                     NaN       NaN     NaN        NaN      0.000000   
25%                     NaN       NaN     NaN        NaN    990.000000   
50%                     NaN       NaN     NaN        NaN   2299.000000   
75%                     NaN       NaN     NaN        NaN   4999.000000   
max                     NaN       NaN     NaN        NaN  16350.000000   

       ab_test vehicle_Type  registration_year gear_box      power_PS  model  \
count    39964        35698       39964.000000    37896  39964.000000  37819   
unique       2            8                NaN        2           NaN    238   
top       test    limousine                NaN  manuell           NaN   golf   
freq     20561        10532                NaN    30718           NaN   3379   
mean       NaN          NaN        2002.675083      NaN    109.948253    NaN   
std        NaN          NaN          35.644961      NaN    194.010425    NaN   
min        NaN          NaN        1910.000000      NaN      0.000000    NaN   
25%        NaN          NaN        1999.000000      NaN     68.000000    NaN   
50%        NaN          NaN        2002.000000      NaN    105.000000    NaN   
75%        NaN          NaN        2006.000000      NaN    141.000000    NaN   
max        NaN          NaN        9000.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_Type       brand  \
count    39964.000000        39964.000000     36260       39964   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     23778        8895   
mean    142286.557902            5.698754       NaN         NaN   
std      16537.626480            3.740147       NaN         NaN   
min      90000.000000            0.000000       NaN         NaN   
25%     150000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_Of_Pictures   postal_Code  \
count              31530                39964         39964.0  39964.000000   
unique                 2                   68             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               27054                 1537             NaN           NaN   
mean                 NaN                  NaN             0.0  50408.937168   
std                  NaN                  NaN             0.0  25620.312782   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  29699.000000   
50%                  NaN                  NaN             0.0  49090.000000   
75%                  NaN                  NaN             0.0  70771.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_Seen  
count                 39964  
unique                33465  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

## Exploring the Odometer & Price column
after finding the columns which need more investigation, first of all, we take a look on two columns 'Odometer' and price.
after a while, we notice that both the columns have some outliers.
to define how to size the outliers, we use the below formula:
    a) if it is greater than Q3+1.5(IQR)
    b) if it is lower than Q1-1.5(IQR)
we should these values which are in this criteria.
for 'price' column, the unforbidden criteria is (0,16350)
for 'odometer_km' column, the criteria is (87500,150000)
finally, We remove the rows step by step which aren't included.
and also we'll have 39964 rows!

In [21]:
auto['date_Crawled'].str[:10].value_counts( normalize=True,dropna=False).head(15)

2016-04-03    0.038184
2016-03-21    0.037534
2016-03-12    0.037384
2016-03-20    0.037309
2016-03-14    0.037158
2016-04-04    0.036733
2016-03-07    0.036283
2016-04-02    0.034706
2016-03-28    0.034631
2016-03-08    0.034231
2016-03-15    0.034006
2016-03-30    0.033981
2016-03-29    0.033830
2016-03-09    0.033805
2016-03-19    0.033580
Name: date_Crawled, dtype: float64

In [28]:
auto['date_Crawled'].str[:10].describe()

count          39964
unique            34
top       2016-04-03
freq            1526
Name: date_Crawled, dtype: object

In [8]:
auto['ad_created'].str[:10].value_counts(normalize=True, dropna=False).head(15)

2016-04-03    0.038460
2016-03-21    0.037659
2016-03-20    0.037309
2016-04-04    0.037259
2016-03-12    0.037234
2016-03-14    0.035607
2016-03-07    0.035132
2016-03-28    0.034606
2016-04-02    0.034306
2016-03-08    0.034206
2016-03-29    0.033931
2016-03-15    0.033805
2016-03-30    0.033780
2016-03-09    0.033680
2016-04-01    0.033180
Name: ad_created, dtype: float64

In [24]:
auto['ad_created'].str[:10].describe()

count          39964
unique            68
top       2016-04-03
freq            1537
Name: ad_created, dtype: object

In [25]:
auto['last_Seen'].str[:10].value_counts( normalize=True,dropna=False).head(15)

2016-04-06    0.209063
2016-04-07    0.122410
2016-04-05    0.118457
2016-03-17    0.029777
2016-04-02    0.025673
2016-04-03    0.025623
2016-03-12    0.025598
2016-03-30    0.025448
2016-04-04    0.025323
2016-03-31    0.024597
2016-04-01    0.024222
2016-03-29    0.023046
2016-03-22    0.022370
2016-03-28    0.022045
2016-03-20    0.022020
Name: last_Seen, dtype: float64

In [26]:
auto['last_Seen'].str[:10].describe()

count          39964
unique            34
top       2016-04-06
freq            8355
Name: last_Seen, dtype: object

### Explain Observation and Analyze the Results
after observing the three columns and see their trends, we can understand some features and make some predictions.
firs of all, for the "data_crawled",
As we see the dates, as we go forward in the dates,the abundance of them has a ascending rate.
second, the 'last_seen',
As we know both columns 'last_seen' & 'data_crawled' are related and there is an algorithm for the ad which posted and the times of showing it. You can see the unique values of thesee columns are same.
third, the 'ad_created' column, 
it has a swinging rate and the special times like weekend night(saturday night) are the busiest time of the week for example, And if You noticed, it has a ascending rate like the two ones too, Overall!




In [9]:
auto['registration_year'].describe()

count    39964.000000
mean      2002.675083
std         35.644961
min       1910.000000
25%       1999.000000
50%       2002.000000
75%       2006.000000
max       9000.000000
Name: registration_year, dtype: float64

### Exploring the 'registration_year'
as you see, there are some unreal and outliers.The year which eBay has just started is 1995 and these data is for until 2006. So the resume should be removed.


In [8]:
auto=auto[auto['registration_year'].between(1995,2006)]
auto['registration_year'].value_counts(normalize=True)

2000    0.111725
1999    0.102391
2005    0.097329
2003    0.091692
2001    0.091477
2004    0.090938
2002    0.085446
2006    0.084799
1998    0.082968
1997    0.069685
1996    0.048718
1995    0.042830
Name: registration_year, dtype: float64

### Removing the outlier values from 'registration_year'
as we mentioned the year which ebay is founded and this dataframe includes until 2006, We set the bound between1995-2006
#### Observing after removing
As you see, there is a yearly low rate in the first years, But after a while in 1998 and after that, It got a jump and spent some shiny years, but after that is has a steady rate, which could have lot of reasons....


In [8]:
auto

date_Crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
32     2016-03-20 05:03:03                               Corsa_mit_TÜV_5.2016   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34     2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
39     2016-03-11 13:58:38                 Mercedes_C_180_Kombi__zu_verkaufen   
40     2016-03-07 14:50:03  VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...   
41     2016-03-10 10:46:08  Passat_3b_1.9_TDI_Highline__angemeldet_mit_tuv...   
44     2016-03-12 10:38:06                                       Zu_verkaufen   
45     2016-03-12 17:39:55                                 Toyota_Starlet_1.3   
46     2016-03-31 10:53:28                                       BMW_mit__Lpg   
...                    ...                                                ...   
49940  2016-03-11 14:52:16                                      Omas_Liebling   
49944  2016-03-15 17:48:16  SAAB_9_3__2_0t_SE__Automatik__Tempomat__Regens...   
49945  2016-04-04 12:54:29                  omega_2_2_facelift_mit_neuem_tuev   
49952  2016-03-21 21:45:33  VW_POLO__9N3_Silber_1_2_Ltr._5_tuerer__96_tsd....   
49954  2016-03-17 19:48:47                                328_Cabrio_Schalter   
49956  2016-03-28 22:39:12        Toyota_Yaris_1.3_VVT_i_Executive_MODEL_2007   
49957  2016-03-10 10:52:01      Bmw_mit_LPG__super_zustand_2_Jahre_Tuev_klima   
49959  2016-03-28 21:53:21           Mercedes_Benz_C_180_T_Kompressor_Classic   
49960  2016-03-25 22:51:55            Ford_KA_zu_verschenken_***Reserviert***   
49962  2016-03-14 17:57:15  Mitsubishi_Space_Star_1__3_L__Bj_2004_Standhei...   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49967  2016-03-12 11:50:17                      VW_Passat_2_0_TDI_comfortline   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                   

In [14]:
auto['brand'].value_counts(normalize=True,dropna=False)

volkswagen        0.206469
opel              0.126984
bmw               0.113161
mercedes_benz     0.094026
audi              0.078696
ford              0.076255
renault           0.058950
peugeot           0.033532
fiat              0.025454
seat              0.018812
mazda             0.017197
nissan            0.014899
citroen           0.012996
smart             0.012314
skoda             0.011488
toyota            0.010519
volvo             0.010232
mitsubishi        0.009298
alfa_romeo        0.007898
honda             0.007755
hyundai           0.006426
kia               0.005637
sonstige_autos    0.005134
suzuki            0.004883
chrysler          0.004667
mini              0.004380
daihatsu          0.002872
chevrolet         0.002441
subaru            0.002298
saab              0.002118
jeep              0.002118
daewoo            0.002010
rover             0.001975
land_rover        0.001651
porsche           0.001580
jaguar            0.001328
lancia            0.001221
d

In [10]:
brand_counts = auto["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

/dataquest/system/env/python3/lib/python3.4/site-packages/ipykernel/__main__.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy



German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [ ]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = auto[auto["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices



Of the top 5 brands, there is a distinct price gap:

Audi, BMW and Mercedes Benz are more expensive
Ford and Opel are less expensive
Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.


In [ ]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_mileage"])

In [ ]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [ ]:

brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

In [ ]:
brand_info["mean_price"] = mean_prices
brand_info


The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.